include/pybind11/detail/internals.h

```diff
@@ -168,15 +173,31 @@ struct override_hash {
     }
 };
 
+using instance_map = std::unordered_multimap<const void *, instance *>;
+
+struct instance_map_shard {
+    std::mutex mutex;
+    instance_map registered_instances;
+    char padding[64 - (sizeof(std::mutex) + sizeof(instance_map)) % 64];
+};
+
 /// Internal data structure used to track registered instances and types.
 /// Whenever binary incompatible changes are made to this structure,
 /// `PYBIND11_INTERNALS_VERSION` must be incremented.
 struct internals {
+#if PYBIND11_INTERNALS_VERSION >= 6
+    std::mutex mutex;
```

Comment:
Would it make sense to protect this with `ifdef Py_GIL_DISABLED?` A free-threaded build can't talk to a non-free-threaded build? And this isn't needed for non-free-threaded builds. (Not sure what's planned for stable ABI going forward, don't want to create an issue if ABI4 supports both)

Reply:
Yeah, this can just be guarded by `Py_GIL_DISABLED`. You are right that a free-threaded build can't talk to a non-free-threaded build.

I had it this way just because it seemed "cleaner" in a sense to have the fields determined solely by `PYBIND11_INTERNALS_VERSION` instead of both `PYBIND11_INTERNALS_VERSION` and `Py_GIL_DISABLED`, but either way should work.

Reply:
It would allow us to continue to increment `PYBIND11_INTERNALS_VERSION` for GIL builds. I'll see what others think, and I can change it if others agree.